OpenLIFUSonicationPlanner/OpenLIFUSonicationPlanner.py

```diff
@@ -152,6 +171,94 @@ def setParameterNode(self, inputParameterNode: Optional[OpenLIFUSonicationPlanne
+            self.ui.ProtocolComboBox.setDisabled(True)
+        else:
+            self.ui.ProtocolComboBox.setEnabled(True)
+            for protocol in dataLogicParameterNode.loaded_protocols.values():
+                # Better to use ID or name here? Showing both for now
+                self.ui.ProtocolComboBox.addItems(["{} (ID: {})".format(protocol.protocol.name,protocol.protocol.id)]) 
+    
+        # Update transducer combo box
+        self.ui.TransducerComboBox.clear()
+        if len(dataLogicParameterNode.loaded_transducers) == 0:
+            self.ui.TransducerComboBox.addItems(["Select a Transducer"]) 
+            self.ui.TransducerComboBox.setDisabled(True)
+        else:
+            self.ui.TransducerComboBox.setEnabled(True)
+            for transducer in dataLogicParameterNode.loaded_transducers.values():
+                transducer_openlifu = transducer.transducer.transducer
+                # Better to use ID or name here? Showing both for now
+                self.ui.TransducerComboBox.addItems(["{} (ID: {})".format(transducer_openlifu.name,transducer_openlifu.id)]) 
+
+        # Update volume combo box 
+        self.ui.VolumeComboBox.clear()  
+        if len(slicer.util.getNodesByClass('vtkMRMLScalarVolumeNode')) == 0:
+            self.ui.VolumeComboBox.addItems(["Select a Volume"])
+            self.ui.VolumeComboBox.setDisabled(True)
+        else:
+            self.ui.VolumeComboBox.setEnabled(True)
+            for volume_node in slicer.util.getNodesByClass('vtkMRMLScalarVolumeNode'):
+                self.ui.VolumeComboBox.addItems(["{} (ID: {})".format(volume_node.GetName(),volume_node.GetID())])
+          
+        # Update target combo box 
+        self.ui.TargetComboBox.clear()  
+        if len(slicer.util.getNodesByClass('vtkMRMLMarkupsFiducialNode')) == 0:
+            self.ui.TargetComboBox.addItems(["Select a Target"])
+            self.ui.TargetComboBox.setDisabled(True)
+        else:
+            self.ui.TargetComboBox.setEnabled(True)
+            for target_node in slicer.util.getNodesByClass('vtkMRMLMarkupsFiducialNode'):
+                self.ui.TargetComboBox.addItems(["{} (ID: {})".format(target_node.GetName(),target_node.GetID())])
+
+    def onDataParameterNodeModified(self,caller, event) -> None:
+        self.updateComboBoxOptions()
+
+    def onPlanClicked(self):
+        print("Run sonication planning")
+
+        dataLogicParameterNode = self.OpenLIFUDataLogic.getParameterNode()
+        activeTransducer = list(dataLogicParameterNode.loaded_transducers.values())[self.ui.TransducerComboBox.currentIndex]
+        activeProtocol = list(dataLogicParameterNode.loaded_protocols.values())[self.ui.ProtocolComboBox.currentIndex]
+        activeVolume = list(slicer.util.getNodesByClass('vtkMRMLScalarVolumeNode'))[self.ui.VolumeComboBox.currentIndex]
+        activeTarget = list(slicer.util.getNodesByClass('vtkMRMLMarkupsFiducialNode'))[self.ui.TargetComboBox.currentIndex]
```

Comment:
here it looks like we are relying on the assumption that `slicer.util.getNodesByClass` will always return the nodes listed in the same order. it's probably true, but this is not an official guarantee so we probably should not do that. better to maintain a mapping between QComboBox indices and mrml nodes. it could be an object of type `Dict[int,slicer.vtkMRMLNode]` that we maintain for each combobox and re-set on every `updateComboBoxOptions`.

then if we do that for volumes and targets we might as well also do it for transducers and protocols.

so `OpenLIFUSonicationPlannerWidget` would maintain some attributes1 like
```
transducer_dict : Dict[int,SlicerOpenLIFUTransducer]
protocol_dict : Dict[int,SlicerOpenLIFUProtocol]
volume_dict : Dict[int,vtkMRMLScalarVolumeNode]
target_dict : Dict[int,vtkMRMLMarkupsFiducialNode]
```

this is just a suggestion but some way or other we should stop assuming that `slicer.util.getNodesByClass` will always return nodes in the same order.

a different approach could be to use the fact that QComboBox can hold "data" in the items: https://doc.qt.io/qt-6/qcombobox.html#currentData-prop
So maybe you could stick the actual nodes in there as the data ... this would be easier if it works

Reply:
I modified the code to store the data within the dropdown items